### ISSUE TITLE
Allow to use self reference (`_`) in the `if` key

### ISSUE THREAD

**Issue by Mingun** [kaitai-io/kaitai_struct#878]
#877 give me an idea, that is will be worth to use `_` (self-reference) in the `if` key:
```yaml
seq:
  - id: predictor
    type: u1
    if: _ == 42
```
This should generate code that use lookahead parsing and then inspects the results and or saves the result or drop it. Currently this can be implemented (with some limitations due to the #709) as
```yaml
seq:
  - id: predictor
    type: u1
    if: predictor_lookahead == 42
instances:
  predictor_lookahead:
    pos: 0 # position actually depends on the enclosing type in the point of application
    type: u1
```

**Comment by KOLANICH** [comment]
No. `if` semantics is whether the field should be parsed at all (i.e. because parsing it in the circumstances it shouldn't be parsed would result in a parse error).

For discarding parse results we have 2 proposals that complement each other:
#88 - for discarding everything except a single field/ins6ance and simplifying the DAG automatically
https://github.com/kaitai-io/kaitai_struct/issues/525#issuecomment-573198726 - for discarding some fields/instances only explicitly

**Comment by Mingun** [comment]
> (i.e. because parsing it in the circumstances it shouldn't be parsed would result in a parse error)

That is the reason for the `lookahead-error` attribute.

> For discarding already parsed results we have 2 proposals that complement each other:

From my point of view both concepts is more harder to understand than just lookahead concept, _especially_ because KSY in step from it. Specifications (in `seq` part) can be parse with the [PEG](https://en.wikipedia.org/wiki/Parsing_expression_grammar) formalism and that formalism have very clear concept of lookahead.

Both of your suggestions are not parsing formalisms, they are just crutches that need to be added to the language describing a kind of parsing formalism in order to achieve the correct behavior. Conceptually, this is worse than my suggestion.

**Comment by KOLANICH** [comment]
>Specifications (in seq part) can be parse with the PEG formalism and that formalism have very clear concept of lookahead.

Packrat grammars are for dealing with ambiguity present in text-based formal languages. That ambiguity is present there to allow people to easier edit such files without calculating sizes and offsets on every edit. Binary grammars are not targeted to be editable by people in `notepad.exe`, they are targeted to be edited withnspecial software (like KS-based tools, after serialization is implemented) they explicitly contain no ambiguity, so **no conflicts** (it's a killer feature of binary grammars, everyone who have written LL/LR/packrat grammars knows it is a great pain in an ass to resolve all the conflicts and keep the spec workable and usable). Packrat grammars are out of scope here, though LL(1) can probably be transpiled into KS (though it makes almost no sense).

**Comment by Mingun** [comment]
PEG grammars have no ambiguity by design. It is not connected to the text formats at all -- it is the formal way to describe structure of any sequences of elements (bytes, characters, even arbitrary tokens, for example, XML tokens). You sometimes just need to look in the another piece of the stream to decide how to parse. Lookahead is just a convenient way to do that when that "another piece" is just at the current cursor position. As you can see, lookahead solves #877 -- _the binary format_ -- that, according to you, has no ambiguity, so lookahead has nothing to do with ambiguity. It just is the one of ways to resolve it, but that not mean that lookahead used can be used only for that.

In #88 you speculate that it would be convenient to hide some parts of the KSY definition from the public API. Lookahead does exactly that, moreover, it hides that even from the format definition, which makes the definition itself cleaner.

**Comment by KOLANICH** [comment]
>PEG grammars have no ambiguity by design

Thank you for the clarification. My knowledge of text-based grammars  is very limited, I didn't have taken a course on them and haven't read the Dragon Book, so I may be wrong on the terminology and understanding (but have read some blogs and Wikipedia and have crafted some LL(1) grammars). Under "ambiguity" I meant here that in a certain state (the position in the stream + counter in the case of `repeat-expr` + already parsed stuff) multiple next states are possible. Completely no lookahead is needed. Even if we introduced a lookahead (I doubt we will, I don't understand the consequences of this currently, except the fact that in grammars for text-based languages conflicts were a plague and it was hard to resolve them, especially if the tools report them as states of internal state machine), it may require a more suitable syntax.

**Comment by Mingun** [comment]
> Under "ambiguity" I meant here that in a certain state (the position in the stream + counter in the case of `repeat-expr` + already parsed stuff) multiple next states are possible. 

You just described ordinary `switch-on` expression.

Lookahead is useful when we can choose between several types in the concrete parse position, but we don't have explicit type selector field, and instead type selector is embedded to types. For example, some format could have numeric field that equals to `0` in the old format (i.e. reserved) and represents some size in the new format. Having such field not in the sub-formats structures can be uncomfortable.

**Comment by KOLANICH** [comment]
>You just described ordinary switch-on expression.

Not quite. `switch-on: expression` requires components of the expression to be already available. By the time the code transpiled from `switch-on` is reached all the ingrindients of the expression must be available. When the transpilation of `switch-on`  is executed, it is precisely known which type goes next, no backtracing is needed and no machinery for it is implemented in KS.

If I understand what `lookahead` means right, we continue parsing one alternative for `n` tokens forward, then, if it parses, then OK, otherwise we revert the state and  parse the next alternative, and so on and so on, and if all the alternatives fail, then c'est la vie, syntax error. It is completely different from how KS works (if I understand, it is like LL(0) with some extensions to understand numbers and strings, but LL(1) can likely be implemented above it with hacks with `_io`) and how binary formats are designed. I am also not sure how it would impact serialization.

>For example, some format could have numeric field that equals to 0 in the old format (i.e. reserved) and represents some size in the new format. Having such field not in the sub-formats structures can be uncomfortable.

```yaml
types:
  c:
    seq:
      - id: value
        type:
          switch-on: version
          cases:
            0: a
            1: b
    instances:
      version:
        pos: 32
        type: u1
```
